// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.tsm";
option java_outer_classname = "TsmProto";
option objc_class_prefix = "TSM";

package tsm;

// The joining service definition.
service TinySocialMedia {
  // Sends a greeting
  rpc joinServer (UserName) returns (UserId) {}
  rpc follow (FollowCommand) returns (GeneralStatus) {}
  rpc unfollow (UnfollowCommand) returns (GeneralStatus) {}
  rpc listUsers (UserId) returns (ListReply) {}
  rpc timeline (stream ClientPost) returns (stream ServerPost) {}
}

// The request message containing the user's name.
message UserName {
  string name = 1;
}

// The response message showing user index
message UserId {
  int64 id = 1;
}


message FollowCommand {
  int64 sender = 1;
  string nameToFollow = 2;
}


message UnfollowCommand {
  int64 sender = 1;
  string nameToUnfollow = 2;
}


enum GStatus
{
    SUCCESS = 0;
    FAILURE_ALREADY_EXISTS = 1;
    FAILURE_NOT_EXISTS = 2;
    FAILURE_INVALID_USERNAME = 3;
    FAILURE_INVALID = 4;
    FAILURE_UNKNOWN = 5;
}

// The response message showing if user is new or not
message GeneralStatus{
  GStatus status = 1; 
}



message ListReply{
  GStatus status = 1;
  repeated string users = 2;
  repeated string followings = 3;
}


message ClientPost {
  int64 sender = 1;
  string content = 2;
}


message ServerPost {
  string content = 1;
  int32 time = 2;
  string senderName = 3;
}